fireant/queries/builder/dataset_blender_query_builder.py

```diff
@@ -0,0 +1,336 @@
+import copy
+
+from pypika import (
+    JoinType,
+    Table,
+    functions as fn,
+)
+from pypika.terms import AggregateFunction
+
+from fireant import (
+    Field,
+    Join,
+)
+from fireant.dataset.modifiers import DimensionModifier
+from fireant.queries import DataSetQueryBuilder
+from fireant.utils import alias_selector
+from ..field_helper import (
+    make_orders_for_dimensions,
+)
+from ..finders import (
+    find_and_replace_reference_dimensions,
+    find_metrics_for_widgets,
+    find_operations_for_widgets,
+    find_share_dimensions,
+)
+from ..sql_transformer import (
+    make_slicer_query_with_totals_and_references,
+)
+
+
+def _get_pypika_field_name_or_alias(pypika_field):
+    """
+    Returns the name or alias, if any, of the provided PyPika Field instance.
+
+    :param pypika_field:
+        A PyPika Field instance.
+    :return: a string.
+    """
+    return pypika_field.alias if pypika_field.alias else pypika_field.name
+
+
+class DataSetBlenderQueryBuilder(DataSetQueryBuilder):
+    """
+    Blended slicer queries consist of widgets, dimensions, filters, orders by and references. At least one or
+    more widgets is required. All others are optional.
+    """
+
+    def __init__(self, dataset, aggregation_func_mapping=None):
+        super(DataSetBlenderQueryBuilder, self).__init__(dataset)
+        self.primary_dataset = dataset.primary_dataset
+        self.secondary_datasets = dataset.secondary_datasets
+        self.all_datasets = [self.primary_dataset, *self.secondary_datasets]
+
+        if aggregation_func_mapping is None:
+            # SUM is used by default when no mapping is found
+            aggregation_func_mapping = {}
+
+        self.aggregation_func_mapping = aggregation_func_mapping
+
+    @property
+    def sql(self):
```

Comment:
this is massive and needs a refactor. This could be a lot simpler if you would leverage the existing query builder code in the data set. Instead of repeating all of the logic, use the `DataSet.query` to build up a query in each of the data sets.

Reply:
@twheys done the refactoring and reused the _DataSet.query_ to build up a query in each of the data sets. Using the sql method of the nested datasets didn't really made it simpler though, given most of the logic we need, either way, is related to the the sub query replacement, custom name changes in metrics/dimensions, deep copying of elements and corner cases (e.g. not applying dimension modifiers in the wrapping query, not applying dimension filters in the wrapping query, use actual aliases for referencing nested fields). I've added three extra tests for cases with references as well. Please see the tests for the DataSetBlender on https://github.com/kayak/fireant/pull/253/files#diff-f4128f84d7aa2c2b06a3c8d8827a4069 once again.